raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,321 @@
+    Token public token; // Only allow RDN?
+
+    // The minimum allowed deposit for a MS
+    uint256 public minimum_deposit;
+
+    // keep track of registered Monitoring Services
+    // A Monitoring Services must deposit to register
+    mapping(address => bool) public registered_monitoring_services;
+
+    // channel_identifier => Struct
+    // Keep track of the rewards per channel
+    mapping(bytes32 => Reward) rewards;
+
+    // Keep track of balances
+    mapping(address => uint) public balances;
+
+    /*
+     *  Structs
+     */
+    struct Reward{
+        // The signature of the reward
+        bytes reward_proof_signature;
+
+        // Nonce of the most recently provided BP
+        uint256 nonce;
+
+        // Address of the Raiden Node that was monitored
+        // This is also the address that has the reward deducted from its deposit
+        address reward_sender_address;
+    }
+
+    /*
+     *  Events
+     */
+
+    //event MonitoringServiceNewDeposit(address sender, address receiver, uint amount);
+    event NewDeposit(address sender, address receiver, uint amount);
+    event RegisteredMonitoringService(address monitoring_service);
+    event NewBalanceProofReceived(
+        bytes reward_proof_signature,
+        uint256 nonce,
+        address ms_address,
+        address raiden_node_address
+    );
+    event RewardClaimed(address ms_address, uint amount, bytes32 channel_id);
+    event MonitoringServiceRegistered(address ms_address);
+    event MonitoringServiceDeregistered(address ms_address);
+    event Withdrawn(address account, uint amount);
+
+    /*
+     *  Modifiers 
+     */
+
+    modifier isMonitor(address _ms_address) {
+        require(registered_monitoring_services[_ms_address]);
+        _;
+    }
+
+    modifier isNotMonitor(address raiden_node_address) {
+        require(!registered_monitoring_services[raiden_node_address]);
+        _;
+    }
+
+    /*
+     *  Constructor
+     */
+
+    /// @notice Set the default values for the smart contract
+    /// @param _token_address The address of the token to use for rewards
+    /// @param _minimum_deposit The minimum value that a MS must deposit to be registered
+    constructor(
+        address _token_address,
+        uint256 _minimum_deposit
+    )
+        public
+    {
+        require(_token_address != 0x0);
+        require(_minimum_deposit > 0);
+        require(contractExists(_token_address));
+
+        minimum_deposit = _minimum_deposit;
+
+        token = Token(_token_address);
+        // Check if the contract is indeed a token contract
+        require(token.totalSupply() > 0);
+    }
+
+    /// @notice Deposit tokens used to either pay MSs or to regsiter as MS
+    /// Can be called by anyone several times and on behalf of other accounts
+    /// @param beneficiary The account benefitting from the deposit
+    /// @param amount The amount of tokens to be depositted
+    function deposit(address beneficiary, uint amount) public {
```

Comment:
As mentioned, we should have 2 functions for MS deposits and user deposits.

Reply:
We actually had this in one of the first versions, but changed it to one function, but I am fine with either.